lib/discovery.js

```diff
@@ -0,0 +1,759 @@
+import net from 'net'
+
+import { TypedEmitter } from 'tiny-typed-emitter'
+import SecretStream from '@hyperswarm/secret-stream'
+import Hyperswarm from 'hyperswarm'
+import { MdnsDiscovery } from 'mdns-sd-discovery'
+import base32 from 'base32.js'
+
+/**
+ * @typedef {Object} DiscoveryEvents
+ * @property {(connection: NoiseSecretStream, info: PeerInfo) => void} connection
+ * @property {(status: { topic: TopicString, mdns: TopicServiceStatus, dht: TopicServiceStatus }) => void} status
+ * @property {(error: Error) => void} error
+ */
+
+/**
+ * @extends {TypedEmitter<DiscoveryEvents>}
+ */
+export class Discovery extends TypedEmitter {
+  #identityKeyPair
+  #topics = new Map()
+  #peers = new Map()
+  #tcp
+
+  /**
+   * @param {Object} options
+   * @param {IdentityKeyPair} options.identityKeyPair
+   * @param {Boolean|Object} options.mdns
+   * @param {Boolean|Object} options.dht
+   */
+  constructor(options) {
+    super()
+
+    const { identityKeyPair, dht = true, mdns = true } = options
+
+    this.#identityKeyPair = identityKeyPair
+
+    if (mdns) {
+      this.#tcp = net.createServer()
+      this.mdns = mdns
+    }
+
+    if (dht) {
+      this.dhtOptions = /** @type {DhtOptions} */ (
+        dht && typeof dht === 'object' ? dht : {}
+      )
+
+      this.dhtOptions.keyPair = identityKeyPair
+      this.dht = new Hyperswarm(this.dhtOptions)
```

Comment:
Stricter typing of this.dht will help catch issues with this.dht being undefined - I think there is a risk of this in later code.

Reply:
made an issue for this https://github.com/digidem/mapeo-core-next/issues/14